lib/email/sender.rb

```diff
@@ -514,5 +474,144 @@ def self.bounceable_reply_address?
+    #    outbound_message_id column on the Post, to survive moves of that
+    #    posts between topics.
+    #
+    # The RFC that covers using "Identification Fields", which are References,
+    # In-Reply-To, Message-ID, et. al. can be found here. It's a good idea to read
+    # this beginning in the area immediately after these quotes, at least to understand
+    # the 3 main headers:
+    #
+    # > The "Message-ID:" field provides a unique message identifier that
+    # > refers to a particular version of a particular message.  The
+    # > uniqueness of the message identifier is guaranteed by the host that
+    # > generates it.
+    #
+    # > ...
+    #
+    # > The "In-Reply-To:" field may be used to identify the message (or
+    # > messages) to which the new message is a reply, while the "References:"
+    # > field may be used to identify a "thread" of conversation.
+    #
+    # https://www.rfc-editor.org/rfc/rfc5322.html#section-3.6.4
+    #
+    # It is a long read, but to understand the decision making process here
+    # you can take a look at:
+    #
+    # https://meta.discourse.org/t/discourse-email-messages-are-incorrectly-threaded/233499
+    def add_experimental_identification_field_headers(topic, post)
+      @message.header["Message-ID"] = Email::MessageIdService.generate_or_use_existing(post)
+
+      if post.post_number > 1
+        ##
+        # Whenever we reply to a post directly _or_ quote a post, a PostReply
+        # record is made, with the reply_post_id referencing the newly created
+        # post, and the post_id referencing the post that was quoted or replied to.
+        referenced_posts = find_referenced_posts(post)
+
+        # TODO (martin) Is it correct to refer to the OP in this case???
+        ##
+        # No referenced posts means that we are just creating a new post not
+        # referring to anything, and as such we should just fall back to using
+        # the OP.
+        if referenced_posts.empty?
+          op_message_id = Email::MessageIdService.generate_or_use_existing(topic.first_post)
+          @message.header["In-Reply-To"] = op_message_id
+          @message.header["References"] = op_message_id
+        else
+          # TODO (martin) Do we need to include the OP Message-ID here????
+          #
+          ##
+          # When referencing _multiple_ posts then we just choose the most recent one
+          # to use for In-Reply-To and References, then also include that post's reply
+          # parents in References as well.
+          most_recent_post_message_id = Email::MessageIdService.generate_or_use_existing(referenced_posts.first)
+          @message.header["In-Reply-To"] = most_recent_post_message_id
+
+          # TODO (martin) Is it correct to include these reply of reply References here???
+          ##
+          # The RFC specifically states that the content of the parent's References
+          # field (in our case all the other post message IDs based on PostReply)
```

Comment:
Is `parent_refrerenced_posts` (spelling?) a chain from the OP through to an immediate parent?

In the RFC the (chosen) parent has a References already, so by construction appending the parent message-id to that makes a complete ordered chain from the OP through the parent.

Here we need to construct that chain from the db. Are we doing that? My mind's too fuzzy right now.